bin/matrixMarket2csv.R

```diff
@@ -17,29 +17,40 @@ opt <- parse_args(OptionParser(option_list=option_list))
 # following columns: experiment_accession, gene_id, cell_id, expression_level
 
 # Read data of the matrix
-readMM(gzfile(opt$matrix_path))->tpm_mtrx
+tpm_mtrx <- readMM(gzfile(opt$matrix_path))
 
 # Read rows (Genes), skipping index row. Is this safe? Is there always a gene name?
-genes_i<-read.table(file=gzfile(opt$genes_path),
-           header = FALSE,
-           col.names = c("index","gene"),
-           colClasses = c("NULL","character"))
+genes_i <- read.table(file = gzfile(opt$genes_path),
+                      header = FALSE,
+                      col.names = c("index", "gene"),
+                      colClasses = c("NULL", "character"))
+
 # Read columns (Cell-id/run)
-runs_j<-read.table(file=gzfile(opt$runs_path),
-           header=FALSE,
-           col.names = c("index","run"),
-           colClasses = c("NULL","character"))
-# Traverse tpm_mtrx object writing sequentially on an object that we write to disk, appending everynow and then.
-
-genes_per_it<-opt$genes_step
-genes_steps<-seq(1,nrow(genes_i),genes_per_it)
-num_genes<-nrow(genes_i)
+runs_j <- read.table(file = gzfile(opt$runs_path),
+                     header =   FALSE,
+                     col.names = c("index","run"),
+                     colClasses = c("NULL","character"))
+
+# Traverse tpm_mtrx object writing sequentially on an object that we write to disk, appending every now and then.
+genes_per_it <- opt$genes_step
+genes_steps <- seq(1, nrow(genes_i), genes_per_it)
+num_genes <- nrow(genes_i)
+
 for(g_i in genes_steps) {
   for(r_j in 1:nrow(runs_j)) {
-      up_to<-min(g_i+genes_per_it-1,num_genes)
-      write.table(data.frame(exp_acc=opt$exp_id,gene_id=genes_i$gene[g_i:up_to],
-                     cell_id=runs_j$run[r_j],expression=tpm_mtrx[g_i:up_to,r_j]),
-                row.names = FALSE, col.names = FALSE, file = opt$output_path, sep = ",",
-                append = TRUE, quote = FALSE)
+    up_to <- min(g_i + genes_per_it - 1, num_genes)
+
+    chunk <- data.frame(exp_acc = opt$exp_id,
+                        gene_id = genes_i$gene[g_i:up_to],
+                        cell_id = runs_j$run[r_j],
+                        expression = tpm_mtrx[g_i:up_to, r_j])
+
```

Comment:
It could well be that the chunk on subsetting for expression > 0 is empty. I would check for this before writing it to disk (knowing prod shared file systems slowness, I would avoid any action of trying to write if it is not needed).

Reply:
If the subset is empty I’d expect `write.table` to be a no-op (other than getting the file descriptor and releasing it). I don’t really think it’s worth it, tbh. I can add it if you’re positive that it will have a noticeable improvement on performance, though.

Reply:
That is normally what takes time on the NFS, getting the file descriptor from the file system. That is why `ls` can sometimes take time on the busy NFS. Add to that that this operation is continuously done (we are writing many chunks per matrix, and that many of them will be empty). So I would rather have that added, is little effort on our side, and on a busy day for the NFS, it will probably pay off.

Reply:
Ok, I addressed the issue you mention. It’s fairly simple but have a look at it, please.